working/0723-static-extensions/feature-specification.md

```diff
@@ -315,165 +453,482 @@ extension.
+from the fully resolved invocation.  These arguments are what are
+needed for subsequent static checking and for the dynamic
+semantics. However, for the purposes of error reporting, it may be
+useful to the user to report errors in terms of the original
+syntactic form.  If `C` is a generic type, then the corresponding type
+arguments for `C` can be obtained simply by substituting the inferred
+type arguments to `E` for the type parameters of `E` in the on-type of
+`E` (which by construction is an instantation of `C`).*
+
+The static type of the constructor invocation in this case is the
+fully instantiated on-type of `E` - that is, the on-type of `E` with
+the inferred type arguments of the fully resolved invocation
+substituted for the type parameters of `E`.
+
+#### Type inference for constructor invocations with explicitly provided type arguments.
+
+Consider an instance creation expression of the form
+`C<TypeArguments>.name(arguments)`, where `E.name(arguments)` is the
+fully resolved invocation of the constructor.
+
+Type inference for such an invocation is done by performing inference
+on the fully resolved invocation `E.name(arguments)` as defined in the
+subsequent section, using `C<TypeArguments>` as the downwards context.
+
+*As above, inference serves to find the type arguments that are
+missing from the fully resolved invocation.  For the purposes of error
+reporting, it may be more useful to the user to report errors in terms
+of the invocation form using `C<TypeArguments>` as given in the
+original program.*
+
+The static type of the constructor invocation in this case is
+`C<TypeArguments>`.
+
+It is a static error if the fully instantiated on-type of `E` - that
+is, the on-type of `E` with the inferred type arguments of the fully
+resolved invocation substituted for the type parameters of `E` - is
+not a subtype of `C<TypeArguments>`.
+
+#### Type inference for fully resolved constructor invocations.
+
+A fully resolved invocation of a constructor declared in an extension
+where either the extension declares no type parameters, or where the
+invocation has type arguments explicitly provided, is subject to no
+further inference to reconstruct the type arguments.  Inference is
+performed as usual on any arguments to the constructor with the
+explicit type arguments used to instantiate the type parameters of the
+extension.
+
+If an extension declares type parameters and no type arguments are
+provided to a fully resolved invocation, then inference is performed
+as follows.  Let `K` be the downwards context of the invocation, and
+let `R` be the on-type of the extension.  Inference is then performed
+in exactly the same manner as an invocation of a static generic method
+in context `K`, the type parameters of which are the type parameters of
+the extension; the return type of which is `R`; and the parameter
+types of which are the declared parameter types of the constructor.
```

Comment:
(There it is: Inference of the invocation of a *function type* with arguments and context type. Except it's done "in exactly the same manner" as for a concrete function with that type, because we haven't defined it by an abstraction of the type.)